### PR TITLE
invert npmignore to include what we want

### PR DESCRIPTION
**Pull request by trusktr**
Closes https://github.com/nasa/openmct/issues/4641

re: https://github.com/nasa/openmct/issues/4511#issuecomment-996220675

### Describe your changes:

instead of excluding a set of files that is more likely to change over time (f.e. ops files, build scripts, ci configs, etc)

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing document](https://github.com/nasa/openmct/blob/master/CONTRIBUTING.md)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/nasa/openmct/pulls) for the same update/change?
* [x] Is this change backwards compatible? For example, developers won't need to change how they are calling the API or how they've extended core plugins such as Tables or Plots.

### Author Checklist

* [x] Changes address original issue?
* [ ] Unit tests included and/or updated with changes?
* [x] Command line build passes?
* [x] Has this been smoke tested?
* [x] Testing instructions included in associated issue?

To test, use `npm pack` to create a tar file, then verify the contents.

### Reviewer Checklist

* [x] Changes appear to address issue?
* [x] Changes appear not to be breaking changes?
* [ ] Appropriate unit tests included?
* [x] Code style and in-line documentation are appropriate?
* [x] Commit messages meet standards?
* [x] Has associated issue been labelled unverified? (only applicable if this PR closes the issue)
* [ ] Has associated issue been labelled bug? (only applicable if this PR is for a bug fix)

### CODE REVIEW

**Comment by trusktr** [comment]
How is it that this increases test coverage?

**Comment by unlikelyzero** [comment]
@trusktr looks good to me. Let's wait until we get the stray .spec.js file removed, update this pr, and then merge. PRs have a habit of getting merged out of order sometimes.

**Comment by unlikelyzero** [comment]
@shefalijoshi please take a look at this with @trusktr . This pattern seems to be far more sustainable.

**Comment by trusktr** [comment]
I believe this PR can be merged separately from fixing that test. The pull request that fixes that test would have to remove the line with `.spec.js` from the npmignore.

File: .npmignore
Comment (unlikelyzero) [inline .npmignore:28]: These two lines can be removed

**Review by unlikelyzero — CHANGES_REQUESTED** [review]
LGTM. Needs Shefali's approval and the two lines removed

File: .npmignore
Comment (trusktr) [reply .npmignore:28]: - [ ] remove lines

**Comment by trusktr** [comment]
@shefalijoshi @unlikelyzero I removed the no-longer-needed line, and also added a `TODO`

File: .npmignore
Comment (trusktr) [inline .npmignore:25]: Wdyt think about this TODO?

Comment (trusktr) [reply .npmignore:25]: There already is a `/src/**/test/` rule ignore anything in `test/` folders. It would make sense to put all non-spec test files like testing utils in such folders, to make it easier to ignore them.

`/src/utils/testing.js` was previously being published to npm.

We can also add one-off rules for specific test files to ignore, but I think it would be easier to have a convention that we just put any test files (apart from the specs that are co-located with the thing they test) in `test/` folders.

**Comment by trusktr** [comment]
To test, use `npm pack` to create a tar file, then verify the contents.

File: .npmignore
Comment (unlikelyzero) [reply .npmignore:25]: @trusktr maybe we could file an issue for that chance and include a link to that issue in this comment?

**Review by unlikelyzero — DISMISSED** [review]
LGTM. I had one suggested change that won't block the merge. Shefali is the release manager so she should give the final 👍

File: .npmignore
Comment (shefalijoshi) [reply .npmignore:25]: We have a naming convention of *Spec*. why can't we use that?

Comment (trusktr) [reply .npmignore:25]: @shefalijoshi Yes, true that, but multiple spec files like to import test-specific utilities from `src/utils/testing.js` so I had to specifically ignore that file because it isn't following a catch-all convention. So what I meant above is that instead we could move that file to `src/test/utils.js` instead, and that way it will fall under the `src/**/test/*` catch-all instead of having a one-off entry in npmignore.

Comment (shefalijoshi) [reply .npmignore:25]: Hmm. Also, for ignoring everything, won't we need to keep updating this when we add new folders?
I prefer ignoring specific files to excluding everything and then including specific items.

Comment (trusktr) [reply .npmignore:25]: > won't we need to keep updating this when we add new folders?

What folders do you imagine we might add? I think keeping things minimal -- src for source and dist for all output -- is the way to go. I believe even the `platform/` folder should go in `src/` because it is source code.

I think it is easier to understand what is included in the published lib (and keeping that simple) rather than having to understand what isn't part of the lib.

The `*Spec.js` lines are the only exception since they're mixed with source files, and they need to be ignore; plus I don't imagine we need to add other types of files into `src/`.

Comment (trusktr) [reply .npmignore:25]: The `.npmignore` that I simply copy/paste into all my JS/TS packages looks like this:

```conf
# ignore everything first (does not ignore special files like README.md, package.json, LICENSE.md, etc)
/**/*

# but include these folders
!/dist/**/*
!/src/**/*

# except for test files in those folders.
/dist/**/*.test.*
/dist/tests/**/*
/src/**/*.test.*
/src/tests/**/*
```

Then from project to project it is really easy to manage. Each project differs and may have different things to ignore, but by convention I can rely on `src/` and `dist/` folders (with test files in them) being the standard, and it stays consistent across projects.

Comment (trusktr) [reply .npmignore:25]: @shefalijoshi Here's a good example!

Recently we split `webpack.config.js` into three files, `webpack.common.js`, `webpack.dev.js`, and `webpack.prod.js`. With the ignore method, we would need to update `npmignore` on any such change (master branch currently does not ignore those files, as well as others, see unpkg.com). With the include method, we wouldn't have to touch `npmignore`.

This is better. I hope I can convince you. 😃

Comment (scottbell) [reply .npmignore:14]: If we do go with this approach, we should add `example` to this too. It has `eventGenerator` and `generator` which a lot of people use to demo OpenMCT

**Comment by scottbell** [comment]
> How is it that this increases test coverage?

Probably a discrepancy between Codecov & Istanbul. Hopefully this:
https://github.com/nasa/openmct/pull/4714
will clear it up.

File: .npmignore
Comment (shefalijoshi) [reply .npmignore:14]: > If we do go with this approach, we should add `example` to this too. It has `eventGenerator` and `generator` which a lot of people use to demo OpenMCT

@trusktr Please include the example, docs folders and Contributing.md, API.md, app.js as well.
What is the procfile used for?

**Review by shefalijoshi — CHANGES_REQUESTED** [review]
Request to include some more files.

File: .npmignore
Comment (trusktr) [reply .npmignore:14]: We want to publish all those things in the package on NPM? I figured people can go to the repo for that stuff. People don't normally look for that stuff in `node_modules`.

Comment (trusktr) [reply .npmignore:14]: Oh I see @scottbell, so some people will want to import from examples.

Comment (trusktr) [reply .npmignore:14]: Do people import `app.js` from `node_modules`. My understanding is we want to obliterate that ( says @akhenry )

**Comment by shefalijoshi** [comment]
@trusktr Please make the suggested changes so we can review and merge this PR. Thanks!

**Comment by jvigliotta** [comment]
@trusktr please look at Shefali's change requests on this.

File: .npmignore
Comment (jvigliotta) [inline .npmignore:16]: can we remove the platform directories? they're no longer in the codebase

**Review by jvigliotta — CHANGES_REQUESTED** [review]
Small comment.

**Comment by jvigliotta** [comment]
@trusktr can you take a look at the review comments?

File: .npmignore
Comment (trusktr) [reply .npmignore]: done

**Comment by trusktr** [comment]
> maybe we could file an issue for that chance and include a link to that issue in this comment?

@unlikelyzero made a new issue #4972

File: .npmignore
Comment (trusktr) [reply .npmignore:14]: > What is the procfile used for?

https://www.google.com/search?q=procfile&oq=procfile&aqs=edge..69i57.870j0j1&sourceid=chrome&ie=UTF-8

Not sure why we need that. Are we using it on some service? If some downstream project needs it, they should have that in their repo instead.

**Comment by trusktr** [comment]
@shefalijoshi wrote:

> Please include the example, docs folders and Contributing.md, API.md, app.js as well.

Do we need to includes all those? Looking for thoughts. Will put a checkbox next to the ones to include

- [ ] app.js: Andrew expressed we should remove, and I agree
- [ ] examples folder: do people consume stuff by importing from `node_modules/openmct/examples`? If not, then they can run examples from the git repo after cloning it
- [ ] docs folder: people can see that in the repo, and on the website, so it doesn't seem we need to ship it to the package registry
- [ ] contributing.md: also visible in the repo, where contributors will be going to contribute anyway.

**Comment by unlikelyzero** [comment]
> Do we need to includes all those? Looking for thoughts. Will put a checkbox next to the ones to include
> 
>  app.js: Andrew expressed we should remove, and I agree
>  examples folder: do people consume stuff by importing from node_modules/openmct/examples? If not, then they can run examples from the git repo after cloning it
>  docs folder: people can see that in the repo, and on the website, so it doesn't seem we need to ship it to the package registry
>  contributing.md: also visible in the repo, where contributors will be going to contribute anyway.

App.js has a removal ticket here: https://github.com/nasa/openmct/issues/4922

**Review by shefalijoshi — APPROVED** [review]
Looks good to me.

File: .npmignore
Comment (jvigliotta) [inline .npmignore:8]: Can you remove this line? We no longer have a platform directory

**Review by jvigliotta — CHANGES_REQUESTED** [review]

**Comment by scottbell** [comment]
>     * [ ]  examples folder: do people consume stuff by importing from `node_modules/openmct/examples`? If not, then they can run examples from the git repo after cloning it
I'm not sure as I don't have data to say one way or another, but I see the Sin Wave, State, and Event generator seem to get a lot of use in openmct deployments.

**Review by unlikelyzero — APPROVED** [review]

**Review by jvigliotta — APPROVED** [review]
LGTM!

**Pull request merged**
